core/internal/filetransfer/file_transfer_s3.go

```diff
@@ -0,0 +1,289 @@
+func (ft *S3FileTransfer) getObjectInputFromTask(
+	bucketName string,
+	objectName string,
+	task *ReferenceArtifactDownloadTask,
+) (*s3.GetObjectInput, error) {
+	var getObjInput *s3.GetObjectInput = &s3.GetObjectInput{
+		Bucket: aws.String(bucketName),
+		Key:    aws.String(objectName),
+	}
+	var getObjAttrsInput *s3.GetObjectAttributesInput = &s3.GetObjectAttributesInput{
+		Bucket:           aws.String(bucketName),
+		Key:              aws.String(objectName),
+		ObjectAttributes: []types.ObjectAttributes{types.ObjectAttributesEtag},
+	}
+
+	versionId, ok := task.VersionIDString()
+	if ok {
+		getObjAttrsInput.VersionId = &versionId
+		getObjInput.VersionId = &versionId
+	}
+
+	objAttrs, err := ft.client.GetObjectAttributes(ft.ctx, getObjAttrsInput)
+	if err != nil {
+		return nil, ft.formatDownloadError("error getting object attributes", err)
+	}
+
+	// If the ETag doesn't match what we have stored, try to find the correct version
+	if trimQuotes(*objAttrs.ETag) != task.Digest {
+		digestMismatchError := fmt.Errorf(
+			"digest/etag mismatch: etag %s does not match expected digest %s",
+			*objAttrs.ETag,
+			task.Digest,
+		)
+		if task.VersionId != nil {
+			return nil, ft.formatDownloadError("", digestMismatchError)
+		}
+		getObjInput, ok = ft.getCorrectObjectVersion(bucketName, objectName, task.Digest, getObjInput)
+		if !ok {
+			return nil, ft.formatDownloadError("", digestMismatchError)
+		}
+	}
+	return getObjInput, nil
+}
+
+// listObjectsWithPrefix returns a list of all objects in the specified bucket
+// that begin with the given prefix
+func (ft *S3FileTransfer) listObjectsWithPrefix(
+	bucketName string,
+	prefix string,
+) ([]*s3.GetObjectInput, error) {
+	var objects []*s3.GetObjectInput
+	params := &s3.ListObjectsV2Input{
+		Bucket: aws.String(bucketName),
+		Prefix: aws.String(prefix),
+	}
+	isTruncated := true
+	for isTruncated {
+		output, err := ft.client.ListObjectsV2(ft.ctx, params)
+		if err != nil {
+			return nil, ft.formatDownloadError("error when retrieving objects", err)
+		}
+
+		for _, object := range output.Contents {
+			objects = append(objects, &s3.GetObjectInput{
+				Bucket: aws.String(bucketName),
+				Key:    object.Key,
+			})
+		}
+
+		isTruncated = *output.IsTruncated
+		if isTruncated {
+			params.ContinuationToken = output.NextContinuationToken
+		}
+	}
+	return objects, nil
+}
+
+// getCorrectObjectVersion attempts to find the version of an object with
+// an ETag that matches the specified digest, and returns the input to
+// retrieve that object
+func (ft *S3FileTransfer) getCorrectObjectVersion(
+	bucketName string,
+	objectName string,
+	digest string,
+	getObjInput *s3.GetObjectInput,
+) (*s3.GetObjectInput, bool) {
+	params := &s3.ListObjectVersionsInput{
+		Bucket: aws.String(bucketName),
+		Prefix: aws.String(objectName),
+	}
+	isTruncated := true
+	for isTruncated {
+		versions, err := ft.client.ListObjectVersions(ft.ctx, params)
+		if err != nil {
+			return nil, false
```

Comment:
I think it would be better to return `digestMismatchError` here instead of returning a bool and having the parent infer that that was the problem